src/TypeScript/TsConfig/TsConfigsLocator.cs

```diff
@@ -0,0 +1,68 @@
+﻿/*
+ * SonarLint for Visual Studio
+ * Copyright (C) 2016-2021 SonarSource SA
+ * mailto:info AT sonarsource DOT com
+ *
+ * This program is free software; you can redistribute it and/or
+ * modify it under the terms of the GNU Lesser General Public
+ * License as published by the Free Software Foundation; either
+ * version 3 of the License, or (at your option) any later version.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
+ * Lesser General Public License for more details.
+ *
+ * You should have received a copy of the GNU Lesser General Public License
+ * along with this program; if not, write to the Free Software Foundation,
+ * Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
+ */
+
+using System;
+using System.Collections.Generic;
+using System.ComponentModel.Composition;
+using System.Linq;
+using Microsoft.VisualStudio;
+using Microsoft.VisualStudio.Shell;
+using Microsoft.VisualStudio.Shell.Interop;
+
+namespace SonarLint.VisualStudio.TypeScript.TsConfig
+{
+    internal interface ITsConfigsLocator
+    {
+        /// <summary>
+        /// Returns all the tsconfig files in the current solution.
+        /// </summary>
+        /// <returns></returns>
+        IEnumerable<string> Locate();
+    }
+
+    [Export(typeof(ITsConfigsLocator))]
+    [PartCreationPolicy(CreationPolicy.Shared)]
+    internal class TsConfigsLocator : ITsConfigsLocator
+    {
+        internal const int MaxNumberOfFiles = 1000;
+
+        private readonly IVsUIShellOpenDocument vsUiShellOpenDocument;
+
+        [ImportingConstructor]
+        public TsConfigsLocator([Import(typeof(SVsServiceProvider))] IServiceProvider serviceProvider)
+        {
+            vsUiShellOpenDocument = serviceProvider.GetService(typeof(SVsUIShellOpenDocument)) as IVsUIShellOpenDocument;
+        }
+
+        public IEnumerable<string> Locate()
+        {
+            var foundFiles = new string[MaxNumberOfFiles];
+
+            var hr = vsUiShellOpenDocument.SearchProjectsForRelativePath(
+                (uint) __VSRELPATHSEARCHFLAGS.RPS_UseAllSearchStrategies,
+                "tsconfig.json",
```

Comment:
I'm not clear on (a) what behaviour we actually want, and (b) whether this achieves it ;-)

For the scenarios we want to support, do we know if the `tsconfig.json` files will always be referenced as project files?
If they don't need to be referenced as project files, will this method find them? (from looking at the docs the answer seems to be "it depends").

Reply:
I tested what VS does if the tsconfig exists on disk but not referenced in the project -- it is still taken into account.
In which case, I believe this method would not work.

So WDYT? Should we change to searching using the file system? I think that's not ideal either, because you could have projects/files that are scattered in the file system and not necessarily under a common root. Moreover, searching by a common root might be incorrect. We could investigate and map VS's behavior in all the cases but I'm not sure if it's worth it. How do you feel about starting with this method and gathering feedback?

Reply:
Yep, I think it's ok to start with this - we have to start somewhere. I don't think searching the file system would be any better; at least this method has some relationship to the solution, even if we're not sure exactly what it is.

The main point was that it made me realise that we don't really know either what we want or what this does so we don't know if the behaviour makes sense, which is not a particularly comfortable place to be. We should definitely focus on testing this with some real-world projects when testing/dogfooding so have a reasonable level of confidence before releasing.